File: Server/OTP.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Security.Cryptography;

namespace Server
{
    class OTP
    {
        // Encrypt with xor plain text and key
        public static string encrypt(string plainText)
        {
            string id, cipherText, key, MAC, result;
            // Split key and id 
            key = nextKey();
            id = key.Substring(1, 1);

            // Find Encrypt key in keys.txt
            key = key.Substring(3, key.Length - 3);

            // Decrypt key 
            //     byte[] k = Encoding.ASCII.GetBytes(key);
            //      key = AES.Decrypt(k);

            // Create cipher Text
            cipherText = xorString(plainText, key);

            // Create MAC
            MAC = createMac(cipherText);

            result = "ID:" + id + "enc:" + cipherText + "MAC:" + MAC;
            return result;
        }

        // Decrypt with xor cipher text and key
        public static string decrypt(string cipherText)
        {
            string key, decrypt, clientMAC, serverMAC, id="", cipher="";
            int i, j=0;

            // Separate MAC,CipherText,ID of key
            for (i = 0; i < cipherText.Length; i++)
            {
                if (cipherText[i] == 'e' && cipherText[i + 1] == 'n')
                {
                    id = cipherText.Substring(3, i - 3);
                    j = i + 4;
                }
                if (cipherText[i] == 'M' && cipherText[i + 1] == 'A')
                {
                    cipher = cipherText.Substring(j, i - j);
                    j = i + 4;
                }
            }
            clientMAC = cipherText.Substring(j, i - j);

            // Find key in kyes.txt
            key = findKEY(id);

            // Decrypt key
         //   byte[] k = Encoding.ASCII.GetBytes(key);
         //   key = AES.Decrypt(k);

            // Decrypt cipher text
            decrypt = xorString(cipher, key);

            // Create MAC with cipher text
            serverMAC = createMac(cipher);

            // Check mac for integrity
            if (serverMAC == clientMAC)
                Console.WriteLine("\nserver mac and client mac match :)) \t");
            else
                Console.WriteLine("\nserver mac and client mac does not match :)) \t");

            return decrypt;
        }

        // XOR two String
        private static string xorString(string str1, string str2)
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < str1.Length; i++)
                sb.Append((char)(str1[i] ^ str2[(i % str2.Length)]));
            String result = sb.ToString();

            return result;
        }

        // Get next key odd key number that until not use from keys.txt file and write to serverSelectedKeys.txt
        public static string nextKey()
        {
            bool flag = false;
            Random rnd = new Random();
            string line = "";
            while (!flag)
            {
                int r = 1+(2*rnd.Next(0, 5));
                r = 3;
                List<string> found = new List<string>();
                using (StreamReader file = new StreamReader($"C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\keys.txt"))
                {
                    while ((line = file.ReadLine()) != null)
                    {
                    //    foreach (string lne in File.ReadLines($"C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\serverSelectedKeys.txt"))
                      //      if (lne == line) continue;

                        if (line.Contains("#3#"))
                        {
                            flag = true;
                            using (StreamWriter writer = new StreamWriter($"C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\serverSelectedKeys.txt"))
                            {
                                writer.WriteLine(line);
                            }
                            break;
                        }
                    }
                }
            }

          //  line = line.Substring(2,line.Length-3);

            return line;
        }

        // Create MAC for integrity
        public static string createMac(string cipherText)
        {
            SHA256 sha256 = SHA256.Create();
            byte[] bytes = Encoding.ASCII.GetBytes(cipherText);
            byte[] MAC = new byte[bytes.Length];
            MAC = sha256.ComputeHash(bytes);

            string mac = Encoding.ASCII.GetString(MAC);

            return mac;
        }

        // Find key that client used with id of keys
        public static string findKEY(string id)
        {
            string line = "";
            id = id + "#";
            List<string> found = new List<string>();
            using (StreamReader file = new StreamReader($"C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\keys.txt"))
            {
                while ((line = file.ReadLine()) != null)
                    if (line.Contains(id)) break;
             }
            line = line.Substring(id.Length+1,line.Length-id.Length-1);
            return line;
        }
    }
}

File: CPP-Client/OTP.cpp
#include "pch.h"
#include "OTP.h"
#include "AES.h"
#include <iostream>
#include <string>

// Encrypt with xor plain text and key (key => is random of even number id key in keys.txt)
std::string OTP::encrypt(std::string plainText) {
	std::string encrypt, MAC, id, result, key, line = "";
	std::ifstream keyFile;
	int j, i;

	// Get random next key in keys.txt(with even id)
	key = nextKey();

	// Split id and key
	for (i = 0; i < key.length(); i++) 
		if (key[i] == '#') {
			id = key.substr(0, i);
			break;
		}

	key = key.substr(i+1, key.length() - i-1);


	// Decrypt key
//	AESENC aes;
//	key = aes.Decryption(key);

	// Encrypt message
	encrypt = xorString(plainText, key);

	// Create MAC
	MAC = createMAC(encrypt);

	//  Message to send server
	result = "ID:" + id + "enc:" + encrypt + "MAC:" + MAC;
	
	return result;
}

// Decrypt with xor cipher text and key
std::string OTP::decrypt(std::string cipherText) {
	std::string decrypt,ServerMAC, ClientMAC, id, cipher, key, line = "";
	std::ifstream keyFile;
	int j,i;

	// Separate MAC,CipherText,ID of key
	for (i = 0; i < cipherText.length(); i++) {
		if (cipherText[i] == 'e' && cipherText[i + 1] == 'n')
		{
			id = cipherText.substr(3, i - 3);
			j = i+4;
		}
		if (cipherText[i] == 'M' && cipherText[i + 1] == 'A')
		{
			cipher = cipherText.substr(j, i - j);
			j = i + 4;
		}
	}
	ServerMAC = cipherText.substr(j, i - j);

	// Find key in keys.txt
	key = findKey(id);

	// Decrypt key
//	AESENC aes;
//	key = aes.Decryption(key);

	// Decrypt response with xor key and cipher text
	decrypt = xorString(cipher, key);

	// Get mac of message
	ClientMAC = createMAC(cipher);

	// Check integerity of message
	if(ClientMAC == ServerMAC)
		std::cout << std::endl <<"server mac and client mac match :)) \t" << std::endl;
	else
		std::cout << std::endl << "server mac and client does not match :||| \t" << std::endl;

	return decrypt;
}

// XOR two String
std::string OTP::xorString(std::string& str1, std::string& str2) {
	// Length of xor output
	int strLength = str1.length() < str2.length() ? str1.length() : str2.length();
	if (str1.length() != str2.length())
		std::cout << "Input sizes do not match! Using the lower one.";
	std::string output;
	
	// Xor char by char
	for (int i = 0; i < strLength; i++)
		output += (str1[i] ^ str2[i]);
	return output;
}

// Get next key even key number (with even id)
std::string OTP::nextKey() {
	std::string key, line = "",lne,rndstr,sharp="#";
	std::ifstream keyFile, keyReadClientFile;
	std::ofstream keyWriteClientFile;
	bool flag = false;
	std::stringstream ss;
	int rnd;
	rnd = createRand();
	rnd = 2;
	ss << rnd;
	rndstr = ss.str();
	rndstr.append("#");
	rndstr.insert(0, "#");
	while (!flag) 
	{
		keyFile.open("C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\keys.txt");
		keyReadClientFile.open("C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\clientSelectedKeys.txt");
		while (getline(keyFile, line)) { 
			//while (getline(keyReadClientFile, lne))if (lne == line) continue;
			if (line.find(rndstr, 0) != std::string::npos) {
				key = line.substr(1, line.length()-1);
				flag = true;

				keyWriteClientFile.open("C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\clientSelectedKeys.txt");
				keyWriteClientFile << line << std::endl;
				keyWriteClientFile.close();
			///	key = line.substr(rnd.length() + 1, line.length() - 1 - rnd.length());
				break;
			}
		}
		keyReadClientFile.close();
		keyFile.close();
	}
	return key;
}

// Find key in keys.txt that use in server
std::string OTP::findKey(std::string id) {
	std::string  key, line = "";
	std::ifstream keyFile;
	id = "#" +id + "#";
	keyFile.open("C:\\Users\\mohsen\\Documents\\Visual Studio 2017\\Projects\\ServerClient\\OneTimePad\\Server\\key\\keys.txt");
	while (getline(keyFile, line)) { // I changed this, see below
		if (line.find(id, 0) != std::string::npos) {
			key = line.substr(id.length(), line.length() - id.length());
		}
	}
	return key;
}

// Mac of input
std::string OTP::createMAC(std::string& str) {
	unsigned char hash[SHA256_DIGEST_LENGTH];
	SHA256_CTX sha256;
	if (!SHA256_Init(&sha256))return false;
	if (!SHA256_Update(&sha256, str.c_str(), str.size()))return false;
	if (!SHA256_Final(hash, &sha256))return false;
	std::stringstream ss;
	for (int i = 0; i < SHA256_DIGEST_LENGTH; i++)
	{
		ss << hash[i];
	}
	return ss.str();
}

// Rand number generator
int OTP::createRand() {
//	return 2 * (rand() % 2);
	return 2;
}